File: apps/web/lib/menu.ts
export interface MenuAccess {
  name: string
  label: string
  icon: string
  path: string
  permission: string
  children?: MenuAccess[]
  accessible?: boolean
}

export interface FeatureAccess {
  export: boolean
  import: boolean
  backup: boolean
  maintenance: boolean
}

export interface MenuAccessResponse {
  menus: MenuAccess[]
  features: FeatureAccess
}

// Client-side menu configuration with icons and metadata
export const menuConfig: Omit<MenuAccess, 'accessible'>[] = [
  {
    name: 'dashboard',
    label: 'Dashboard',
    icon: 'Home',
    path: '/dashboard',
    permission: 'menu.dashboard'
  },
  {
    name: 'admin-panel',
    label: 'Admin Panel',
    icon: 'Shield',
    path: '/admin',
    permission: 'menu.admin-panel'
  },
  {
    name: 'analytics',
    label: 'Analytics',
    icon: 'BarChart3',
    path: '/analytics',
    permission: 'menu.analytics'
  },
  {
    name: 'reports',
    label: 'Reports',
    icon: 'FileText',
    path: '/reports',
    permission: 'menu.reports'
  },
  {
    name: 'admin',
    label: 'Administration',
    icon: 'Settings',
    path: '/admin',
    permission: 'menu.admin',
    children: [
      {
        name: 'users',
        label: 'User Management',
        icon: 'Users',
        path: '/admin/users',
        permission: 'menu.users'
      },
      {
        name: 'roles',
        label: 'Role Management',
        icon: 'Shield',
        path: '/admin/roles',
        permission: 'menu.roles'
      },
      {
        name: 'audit',
        label: 'Audit Logs',
        icon: 'Clock',
        path: '/admin/audit',
        permission: 'menu.audit'
      }
    ]
  },
  {
    name: 'billing',
    label: 'Billing',
    icon: 'CreditCard',
    path: '/billing',
    permission: 'menu.billing'
  },
  {
    name: 'support',
    label: 'Support',
    icon: 'HelpCircle',
    path: '/support',
    permission: 'menu.support'
  },
  {
    name: 'settings',
    label: 'Settings',
    icon: 'Cog',
    path: '/settings',
    permission: 'menu.settings'
  }
]

// Role-based access documentation for UI
export const roleAccessMap = {
  admin: {
    name: 'Administrator',
    description: 'Full system access including admin panel',
    color: 'red',
    menus: ['Dashboard', 'Analytics', 'Reports', 'Administration', 'Admin Panel', 'Billing', 'Support', 'Settings']
  },
  manager: {
    name: 'Manager', 
    description: 'Business analytics access',
    color: 'blue',
    menus: ['Dashboard', 'Analytics', 'Reports', 'Billing']
  },
  editor: {
    name: 'Editor',
    description: 'Content management access', 
    color: 'green',
    menus: ['Dashboard', 'User Management', 'Support']
  },
  viewer: {
    name: 'Viewer',
    description: 'Read-only access',
    color: 'gray',
    menus: ['Dashboard', 'Analytics', 'Reports']
  },
  support: {
    name: 'Support Staff',
    description: 'User assistance access',
    color: 'purple',
    menus: ['Dashboard', 'Support', 'User Management']
  },
  user: {
    name: 'User',
    description: 'Basic dashboard access',
    color: 'yellow', 
    menus: ['Dashboard']
  }
}

File: apps/backend/utils/menu.go
package utils

import (
	"backend/models"
)

// MenuAccess represents menu access configuration
type MenuAccess struct {
	Name       string       `json:"name"`
	Label      string       `json:"label"`
	Icon       string       `json:"icon"`
	Path       string       `json:"path"`
	Permission string       `json:"permission"`
	Children   []MenuAccess `json:"children,omitempty"`
	Accessible bool         `json:"accessible"`
}

// GetUserMenuAccess returns accessible menu items based on user permissions
func GetUserMenuAccess(user *models.User) []MenuAccess {
	// Define all menu items with their required permissions
	allMenus := []MenuAccess{
		{
			Name:       "dashboard",
			Label:      "Dashboard",
			Icon:       "dashboard",
			Path:       "/dashboard",
			Permission: "menu.dashboard",
		},
		{
			Name:       "admin-panel",
			Label:      "Admin Panel",
			Icon:       "admin_panel_settings",
			Path:       "/admin",
			Permission: "menu.admin-panel",
		},
		{
			Name:       "analytics",
			Label:      "Analytics",
			Icon:       "analytics",
			Path:       "/analytics",
			Permission: "menu.analytics",
		},
		{
			Name:       "reports",
			Label:      "Reports",
			Icon:       "report",
			Path:       "/reports",
			Permission: "menu.reports",
		},
		{
			Name:       "admin",
			Label:      "Administration",
			Icon:       "admin_panel_settings",
			Path:       "/admin",
			Permission: "menu.admin",
			Children: []MenuAccess{
				{
					Name:       "users",
					Label:      "User Management",
					Icon:       "people",
					Path:       "/admin/users",
					Permission: "menu.users",
				},
				{
					Name:       "roles",
					Label:      "Role Management",
					Icon:       "security",
					Path:       "/admin/roles",
					Permission: "menu.roles",
				},
				{
					Name:       "audit",
					Label:      "Audit Logs",
					Icon:       "history",
					Path:       "/admin/audit",
					Permission: "menu.audit",
				},
			},
		},
		{
			Name:       "billing",
			Label:      "Billing",
			Icon:       "receipt",
			Path:       "/billing",
			Permission: "menu.billing",
		},
		{
			Name:       "support",
			Label:      "Support",
			Icon:       "support_agent",
			Path:       "/support",
			Permission: "menu.support",
		},
		{
			Name:       "settings",
			Label:      "Settings",
			Icon:       "settings",
			Path:       "/settings",
			Permission: "menu.settings",
		},
	}

	// Filter menus based on user permissions
	return filterMenusByPermissions(allMenus, user)
}

// filterMenusByPermissions recursively filters menu items based on user permissions
func filterMenusByPermissions(menus []MenuAccess, user *models.User) []MenuAccess {
	var accessibleMenus []MenuAccess

	for _, menu := range menus {
		// Check if user has permission for this menu
		if hasMenuPermission(user, menu.Permission) {
			menu.Accessible = true

			// Filter children if they exist
			if len(menu.Children) > 0 {
				menu.Children = filterMenusByPermissions(menu.Children, user)
			}

			accessibleMenus = append(accessibleMenus, menu)
		}
	}

	return accessibleMenus
}

// hasMenuPermission checks if user has specific menu permission
func hasMenuPermission(user *models.User, permission string) bool {
	for _, role := range user.Roles {
		for _, perm := range role.Permissions {
			if perm.Name == permission {
				return true
			}
		}
	}
	return false
}

// GetUserFeatureAccess returns accessible features based on user permissions
func GetUserFeatureAccess(user *models.User) map[string]bool {
	features := map[string]bool{
		"export":      hasMenuPermission(user, "feature.export"),
		"import":      hasMenuPermission(user, "feature.import"),
		"backup":      hasMenuPermission(user, "feature.backup"),
		"maintenance": hasMenuPermission(user, "feature.maintenance"),
	}

	return features
}

// RoleMenuMap returns predefined menu access for each role (for documentation)
func RoleMenuMap() map[string][]string {
	return map[string][]string{
		"admin": {
			"menu.dashboard", "menu.analytics", "menu.reports", "menu.admin",
			"menu.users", "menu.roles", "menu.audit", "menu.billing",
			"menu.support", "menu.settings",
		},
		"manager": {
			"menu.dashboard", "menu.analytics", "menu.reports", "menu.billing",
		},
		"editor": {
			"menu.dashboard", "menu.users", "menu.support",
		},
		"viewer": {
			"menu.dashboard", "menu.analytics", "menu.reports",
		},
		"support": {
			"menu.dashboard", "menu.support", "menu.users",
		},
		"user": {
			"menu.dashboard",
		},
	}
}
